File: viewmodel/src/main/java/com/senyk/volodymyr/schedulesapp/viewmodel/helpers/SingleEventLiveData.java
package com.senyk.volodymyr.schedulesapp.viewmodel.helpers;

import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

import java.util.concurrent.atomic.AtomicBoolean;

public class SingleEventLiveData<T> extends MutableLiveData<T> {
    private AtomicBoolean pending = new AtomicBoolean(false);

    @Override
    public void observe(@NonNull LifecycleOwner owner, @NonNull final Observer<? super T> observer) {
        super.observe(owner, data -> {
            if (SingleEventLiveData.this.pending.compareAndSet(true, false)) {
                observer.onChanged(data);
            }
        });

    }

    @Override
    public void setValue(T value) {
        this.pending.set(true);
        super.setValue(value);
    }

}

File: core/src/main/java/com/senykvolodymyr/core/livedata/SingleEventLiveData.kt
package com.senykvolodymyr.core.livedata

import androidx.annotation.AnyThread
import androidx.annotation.MainThread
import androidx.annotation.Nullable
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import java.util.concurrent.atomic.AtomicBoolean

class SingleEventLiveData<T> : MutableLiveData<T>() {

    private val alreadySent = AtomicBoolean(true)

    @MainThread
    override fun observe(owner: LifecycleOwner, observer: Observer<in T>) {
        super.observe(owner, { value: T ->
            if (alreadySent.compareAndSet(false, true)) {
                observer.onChanged(value)
            }
        })
    }

    @MainThread
    override fun setValue(@Nullable value: T?) {
        alreadySent.set(false)
        super.setValue(value)
    }

    @AnyThread
    override fun postValue(@Nullable value: T?) {
        alreadySent.set(false)
        super.postValue(value)
    }

    @MainThread
    fun call() {
        value = null
    }
}
